### PR TITLE
[Console] Document console completion

### LINKED ISSUE #15947: [Console] Bash completion integration
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#42251
| PR author(s) | @wouterj
| Merged in    | 5.4

### PR DESCRIPTION
**Pull request by wouterj**
Fixes #15947

I couldn't resist rewriting the Console document a bit, to structure it more like "Usage" and "Implementing a custom command". This created a nice place to talk about setting up completion.

### CODE REVIEW

File: console.rst [hunk 1]
```diff
@@ -9,15 +9,90 @@ The Symfony framework provides lots of commands through the ``bin/console`` scri
...
+    $ php bin/console completion bash | sudo tee /etc/bash_completion.d/console-events-terminate
+    # after the installation, restart the shell
+
+Now you are all set to use the auto completion for all Symfony Console
+applications on your computer. By default, you can get a list of complete
+options by pressing the Tab key.
+
+.. tip::
+
+    Many PHP tools are build using the Symfony Console component (e.g.
```
Comment (javiereguiluz) [inline console.rst]: `are build` -> `are built`

File: console/input.rst [hunk 2]
```diff
@@ -308,4 +308,90 @@ The above code can be simplified as follows because ``false !== null``::
...
+                // the value the user already typed, e.g. when typing "app:greet Fa" before
+                // pressing Tab, this will contain "Fa"
+                $currentValue = $input->getCompletionValue();
+
+                // get the list of username names from somewhere (e.g. the database)
+                // you may use $currentValue to filter down the names
+                $availableUsernames = ...;
+
+                // then add the retrieved names as suggested values
+                $suggestions->suggestValues($availableUsernames);
```
Comment (javiereguiluz) [inline console/input.rst:348]: I miss some comments about the limits of suggestions. Can I return an unlimited number here? Is there a hard limit or a soft recommendation? Thanks!

File: page_creation.rst [hunk 3]
```diff
@@ -195,6 +195,13 @@ the debugging routes in the next section.
 
 You'll learn about many more commands as you continue!
 
+.. tip::
+
+    If you are using the Bash shell, you can set-up completion support.
```
Comment (javiereguiluz) [inline page_creation.rst]: I'm not entirely sure, but I think `set-up` should be `set up`

**Review by javiereguiluz — APPROVED** [review]
Wouter, thanks for adding the docs of this lovely feature!

File: console/input.rst (hunk 2 above)
Comment (wouterj) [reply console/input.rst:348]: It's unlimited (or at least massive, the shell script automatically filters the suggestions and shows a message like "Display all 3197 possibilities? (y or no)" when there are many suggestions).

I've updated the tip below this example a little bit to better highlight it.

**Comment by GromNaN** [comment]
Thank you @wouterj. Do you think this page should be updated for 6.1 with the new features: fish completion https://github.com/symfony/symfony/pull/43641 and configuration in option/argument définition https://github.com/symfony/symfony/pull/44948

File: console/input.rst [hunk 4]
```diff
@@ -308,4 +308,91 @@ The above code can be simplified as follows because ``false !== null``::
...
+            $tester = new CommandCompletionTester($application->get('app:greet'));
+
+            // complete the input without any existing input (the empty string represents
+            // the position of the cursor)
+            $suggestions = $tester->complete(['']);
+            $this->assertSame(['Fabien', 'Fabrice', 'Wouter'], $suggestions);
+
+            // complete the input with "Fa" as input
+            $suggestions = $tester->complete(['Fa']);
+            $this->assertSame(['Fabien', 'Fabrice'], $suggestions);
```
Comment (GromNaN) [inline console/input.rst:394]: Users will think they need to filter the values themself, while this is done by the completion tool.
I think `getCompletionValue` usage is rare and reserved to cases when the list of values is too long (like the list of all packages on packagist https://github.com/composer/composer/pull/10320#discussion_r760258324)

**Comment by wouterj** [comment]
> Do you think this page should be updated for 6.1 with the new features

Yes, see https://github.com/symfony/symfony-docs/pull/16724 (would be great to get your review there)

**Pull request merged**